File: iOS/PanoAudioChat/PanoAudioChatSDK/Source/Services/PanoClientService+OpenApi.m
//
//  PanoClientService+OpenApi.m
//  PanoAudioChatSDK
//
//  Created by pano on 2021/7/15.
//

#import "PanoClientService+OpenApi.h"
#import "PanoClientService+App.h"

@implementation PanoClientService (OpenApi)

- (void)joinWithConfigWithCompletion:(PanoCompletionBlock)block {
    self.joinBlock = block;
    PanoClientConfig *config = self.config;
    self.userName = config.userName;
    self.roomId = config.roomId;
    self.userId = config.userId;
    [self savePreferences];
    [self fetchTokenWithCompletionHandler:^(NSString * _Nonnull token, UInt64 hostUserId, NSError * _Nonnull error) {
        dispatch_async(dispatch_get_main_queue(), ^{
            if (error) {
                block(error);
                return;
            }
            [self createEngineKit];
            PanoRtcChannelConfig *chanelConfig = [[PanoRtcChannelConfig alloc] init];
            chanelConfig.mode = kPanoChannelMeeting;
            chanelConfig.subscribeAudioAll = true;
            chanelConfig.userName = config.userName;
            PanoResult res = [self.engineKit joinChannelWithToken:token channelId:config.roomId userId:config.userId config:chanelConfig];
            if (res == kPanoResultOK) {
                self.joined = true;
                [self.userService onUserJoinIndication:config.userId withName:config.userName];
                UIApplication.sharedApplication.idleTimerDisabled = true;
                self.hostUserId = hostUserId;
            } else {
                block([NSError errorWithDomain:NSCocoaErrorDomain code:res userInfo:nil]);
            }
        });
    }];
}

- (void)leaveAudioRoom {
    if (!self.joined) {
        return;
    }
    self.joinBlock = nil;
    self.joined = false;
    NSTimeInterval after = 0.0;
    if (self.config.userMode == PanoAnchor) {
        /**
           关闭房间有两种方式，
            第一： 通过发送广播消息，然后大家离开会议
            第二：调用RestfulApi close channel
         */
        [self.signalService sendCloseRoomMsg];
        after = 1.5;
    }
    [self uploadlogsAutomatically];
    [self.audioService stopAudioDump];
    [self.playerService unLoadMusicResouces];
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(after * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [self destroyEngineKit];
        UIApplication.sharedApplication.idleTimerDisabled = false;
    });
}

- (UIViewController *)settingViewController {
    NSBundle *bundle = [NSBundle bundleForClass:[self class]];
    UIStoryboard *sb = [UIStoryboard storyboardWithName:@"Setting" bundle: bundle];
    UIViewController *vc = [sb instantiateViewControllerWithIdentifier:@"setting"];
    UINavigationController *navi = [[UINavigationController alloc] initWithRootViewController:vc];
    navi.modalPresentationStyle = UIModalPresentationFullScreen;
    return navi;
}

- (BOOL)sendMessage:(NSDictionary<NSString *,id> *)msg toUser:(UInt64)userId {
    if (self.rtmState == kPanoMessageServiceUnavailable) {
        return false;
    }
    NSError *error = nil;
    NSData *data = [NSJSONSerialization dataWithJSONObject:msg options:0 error:&error];
    if (error) {
        return false;
    }
    PanoResult result = [self.engineKit.messageService sendMessageToUser:userId data:data];
    return result == kPanoResultOK;
}

- (BOOL)broadcastMessage:(NSDictionary<NSString *,id> *)msg {
    if (self.rtmState == kPanoMessageServiceUnavailable) {
        return false;
    }
    NSError *error = nil;
    NSData *data = [NSJSONSerialization dataWithJSONObject:msg options:0 error:&error];
    if (error) {
        return false;
    }
    PanoResult result = [self.engineKit.messageService broadcastMessage:data sendBack:true];
    return result == kPanoResultOK;
}

- (BOOL)setProperty:(id)value forKey:(NSString *)key {
    if (self.rtmState == kPanoMessageServiceUnavailable) {
        return false;
    }
    NSData *data = nil;
    if (value) {
        NSError *error = nil;
        data = [NSJSONSerialization dataWithJSONObject:value options:0 error:&error];
    }
    PanoResult result = [self.engineKit.messageService setProperty:key value:data];
    return result == kPanoResultOK;
}

@end

File: iOS/PanoAudioChat/PanoAudioChat/Classes/Service/PanoClientService+OpenApi.swift
//
//  PanoClientService+OpenApi.swift
//  PanoAudioChat
//
//  Created by wenhua.zhang on 2021/6/11.
//

import Foundation

extension PanoClientService: PanoSignalInterface {
    
    func joinAudioRoom(config: PanoClientConfig, completion: @escaping (Bool) -> Void) {
        self.config = config;
        self.joinCompletion = completion
        PanoClientService.userName = config.userName
        PanoClientService.roomId = config.roomId
        PanoClientService.userId = config.userId
        PanoClientService.savePreferences()
        fetchTokenWithCompletionHandler { (token, hostUserId , error) in
            guard error == nil ,
                  let token = token,
                  let hostUserId = hostUserId,
                  token.count > 0 else {
                completion(false)
                return
            }
            self.createEngineKit()
             
            let chanelConfig = PanoRtcChannelConfig()
            chanelConfig.mode = .channelMeeting
            chanelConfig.subscribeAudioAll = true
            chanelConfig.userName = config.userName
            let res = self.engineKit?.joinChannel(withToken: token, channelId: config.roomId, userId: config.userId, config: chanelConfig)
            if res == .OK {
                print("joinChannel begin", NSDate())
                PanoClientService.joined = true
                let userService = PanoServiceManager.service(type: .user) as? PanoUserService;
                userService?.onUserJoinIndication(config.userId, withName: config.userName)
                UIApplication.shared.isIdleTimerDisabled = true
                PanoClientService.hostUserId = UInt64(Int(hostUserId)!)
            } else {
                completion(false)
            }
        }
    }
    
    @objc func leaveAudioRoom() {
        if !PanoClientService.joined {
            return
        }
        self.joinCompletion = nil
        PanoClientService.joined = false
        var after = 0.0;
        if config.userMode == .anchor {
            /**
               关闭房间有两种方式，
                第一： 通过发送广播消息，然后大家离开会议
                第二：调用RestfulApi close channel
             */
            _ = self.signalService?.sendCloseRoomMsg()
            
            // self.closeChannelWithHandler()
            after = 1.5;
        }
        self.uploadlogsAutomatically()
        PanoAudioService.service()?.stopAudioDump()
        PanoPlayerService.service()?.unLoadMusicResouces()
        engineKit?.whiteboardEngine().setDelegate(nil)
        PanoServiceManager.uninit()
        DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + after) {
            self.destroyEngineKit()
            UIApplication.shared.isIdleTimerDisabled = false
            
        }
    }
    
    func sendMessage(msg: Dictionary<String, Any>, toUser: UInt64) -> Bool {
        print("rtmState->", rtmState.rawValue);
        let data = try! JSONSerialization.data(withJSONObject: msg, options: JSONSerialization.WritingOptions(rawValue: 0))
        let result = engineKit?.messageService.send(toUser: toUser, data: data)
        print("sendMessage result->", result?.rawValue)
        return result == .OK
    }
    
    func broadcastMessage(msg: Dictionary<String, Any>) -> Bool {
        print("rtmState->", rtmState.rawValue);
        let data = try! JSONSerialization.data(withJSONObject: msg, options: JSONSerialization.WritingOptions(rawValue: 0))
        let result = engineKit?.messageService.broadcast(data, sendBack: true)
        print("broadcastMessage result->", result?.rawValue)
        return result == .OK
    }
    
    func setProperty(value: Any? = nil, for key: String) -> Bool {
        var data: Data = Data();
        if let d = value {
            data = try! JSONSerialization.data(withJSONObject: d, options: JSONSerialization.WritingOptions(rawValue: 0))
        }
        let result = engineKit?.messageService.setProperty(key, value: data)
        return result == .OK
    }
    
    func addDelegate(delegate: PanoClientDelegate) {
        delegates.addDelegate(delegate);
    }
    
    /// 移除回调
    func removeDelegate(delegate: PanoClientDelegate) {
        delegates.removeDelegate(delegate);
    }

}
